lib/setup-sandbox.js

```diff
@@ -28,7 +28,24 @@ const {
 	getOwnPropertyDescriptor: localReflectGetOwnPropertyDescriptor
 } = localReflect;
 
+const speciesSymbol = Symbol.species;
+const globalPromise = global.Promise;
+class localPromise extends globalPromise {}
+
+const blockPromiseSpecies = (p, Class) => {
+	if (p instanceof Class && p.constructor[speciesSymbol] !== Class) {
+		throw new Error('Sandbox escape attempt blocked');
+	}
+};
+
+const globalPromiseThen = globalPromise.prototype.then;
+globalPromise.prototype.then = function then(onFulfilled, onRejected) {
```

Comment:
this is unfortunately polluting the global prototype. but until we replace vm2, I can't find an alternative way to do this.